blog/DS-1311/RDS-Encryption-At-Rest.md

```diff
@@ -0,0 +1,391 @@
+	}); err != nil {
+		return Instance{}, err
+	}
+
+	if err := s.waitForDBStatus(instanceName, readyFunc); err != nil {
+		return Instance{}, err
+	}
+
+	// DB instance has to be bounced now, but we need to wait until
+	// parameter group status == pending-reboot before doing so ...
+	readyFunc2 := func(i Instance) bool {
+		ok := i.ParGroupStatus == pendingReboot
+		if !ok {
+			s.log.Printf("... ModifyInstance: [%24s] waiting for %s group status change, want: %s got: %s", instanceName, dbParGroupName, pendingReboot, i.ParGroupStatus)
+		}
+		return ok
+	}
+	if err := s.waitForDBStatus(instanceName, readyFunc2); err != nil {
+		return Instance{}, err
+	}
+
+	// reboot (no failover as per doc)
+	s.log.Printf("... ModifyInstance: [%24s] Rebooting, for DBParameterGroupName %q to take effect", instanceName, dbParGroupName)
+	if err := s.Reboot(instanceName, false); err != nil {
+		return Instance{}, err
+	}
+
+	if err := s.waitForDBStatus(instanceName, readyFunc); err != nil {
+		return Instance{}, err
+	}
+
+	return s.Describe(instanceName)
+}
+```
+
+In the code snippet above, we are calling `waitForDBStatus` twice: the first pass is to ensure the newly restored instance is in `Available` state before doing any changes, and the second time to wait for the reboot to happen before returning back to the caller.  The reboot is needed for `DBParameterGroupName` change to take effect.  If you need the code for `waitForDBStatus`, it's available here: [waitForDBStatus source code](https://github.com/InVisionApp/ds-blog/blob/master/blog/DS-1311/code/wait_for_dbstatus.go)
+
+## Recreate Read Replicas
+At this point, the instance is restored from the encrypted snapshot and all if it's attributes match the source instance it's replacing (we'll go over how to validate this in the following section).  The next step is recreate all of it's read-replicas if any exist.  This task can require substantial custom automation depending on how these read-replicas were originally setup.  For example, we had to redo and automate the following:
+
+1. re-create any DB users that were created directly on the original replica
+2. re-set [binlog retention hours](https://docs.aws.amazon.com/AmazonRDS/latest/UserGuide/mysql_rds_show_configuration.html) that were configured on the original replica
+
+Lets go over these in order:
+
+```go
+func (s *SDK) reCreateReplica(master, copyFrom Instance, name string, binlogRetention int, rootPass string) error {
+	for _, replica := range master.RDSDBInstance.ReadReplicaDBInstanceIdentifiers {
+		if *replica == name {
+			s.log.Printf("... reCreateReplica: [%24s] %q replica already exists", master.Name, name)
+			return s.reCreateReplicaFinalize(master, copyFrom, name, binlogRetention, rootPass)
+		}
+	}
+
+	replicaInput := &rds.CreateDBInstanceReadReplicaInput{
+		AutoMinorVersionUpgrade: copyFrom.RDSDBInstance.AutoMinorVersionUpgrade,
+		// AvailabilityZone:            copyFrom.RDSDBInstance.AvailabilityZone,
```

Comment:
Why is this commented out?

Reply:
it's set later in the code only when copyFrom.RDSDBInstance.MultiAZ is not true - I'll remove it to reduce the noise